File: test/javascript/elixir/bitwise_test.mjs
import {
  assert,
  assertBoxedError,
  defineGlobalErlangAndElixirModules,
} from "../support/helpers.mjs";

import Elixir_Bitwise from "../../../assets/js/elixir/bitwise.mjs";
import Interpreter from "../../../assets/js/interpreter.mjs";
import Type from "../../../assets/js/type.mjs";

defineGlobalErlangAndElixirModules();

describe("Elixir_Bitwise", () => {
  describe("band/2 and &&&/2", () => {
    const band = Elixir_Bitwise["band/2"];
    const operator = Elixir_Bitwise["&&&/2"];

    it("calculates bitwise AND", () => {
      const result = band(Type.integer(9), Type.integer(3));
      const expected = Type.integer(1);

      assert.deepStrictEqual(result, expected);
    });

    it("handles negative operands", () => {
      const result = operator(Type.integer(-5), Type.integer(12));
      const expected = Type.integer(8n & -5n);

      assert.deepStrictEqual(result, expected);
    });

    it("raises ArithmeticError when an operand is not an integer", () => {
      const left = Type.atom("foo");
      const right = Type.integer(1);

      assertBoxedError(
        () => band(left, right),
        "ArithmeticError",
        "bad argument in arithmetic expression: Bitwise.band(:foo, 1)",
      );
    });
  });

  describe("bor/2 and |||/2", () => {
    const bor = Elixir_Bitwise["bor/2"];
    const operator = Elixir_Bitwise["|||/2"];

    it("calculates bitwise OR", () => {
      const result = bor(Type.integer(9), Type.integer(3));
      const expected = Type.integer(11);

      assert.deepStrictEqual(result, expected);
    });

    it("supports operator form", () => {
      const result = operator(Type.integer(-5), Type.integer(12));
      const expected = Type.integer(-5n | 12n);

      assert.deepStrictEqual(result, expected);
    });
  });

  describe("bxor/2 and ^^^/2", () => {
    const bxor = Elixir_Bitwise["bxor/2"];
    const operator = Elixir_Bitwise["^^^/2"];

    it("calculates bitwise XOR", () => {
      const result = bxor(Type.integer(9), Type.integer(3));
      const expected = Type.integer(10);

      assert.deepStrictEqual(result, expected);
    });

    it("supports operator form", () => {
      const result = operator(Type.integer(-5), Type.integer(12));
      const expected = Type.integer(-5n ^ 12n);

      assert.deepStrictEqual(result, expected);
    });
  });

  describe("bnot/1 and ~~~/1", () => {
    const bnot = Elixir_Bitwise["bnot/1"];
    const operator = Elixir_Bitwise["~~~/1"];

    it("calculates bitwise NOT", () => {
      const result = bnot(Type.integer(2));
      const expected = Type.integer(-3);

      assert.deepStrictEqual(result, expected);
    });

    it("supports operator form", () => {
      const result = operator(Type.integer(-1));
      const expected = Type.integer(~-1n);

      assert.deepStrictEqual(result, expected);
    });
  });

  describe("bsl/2 and <<</2", () => {
    const bsl = Elixir_Bitwise["bsl/2"];
    const operator = Elixir_Bitwise["<<</2"];

    it("shifts left for positive shift counts", () => {
      const result = bsl(Type.integer(1), Type.integer(3));
      const expected = Type.integer(8);

      assert.deepStrictEqual(result, expected);
    });

    it("converts negative shift counts to right shifts", () => {
      const result = bsl(Type.integer(1), Type.integer(-2));
      const expected = Type.integer(0);

      assert.deepStrictEqual(result, expected);
    });

    it("supports operator form", () => {
      const result = operator(Type.integer(-1), Type.integer(2));
      const expected = Type.integer(-1n << 2n);

      assert.deepStrictEqual(result, expected);
    });
  });

  describe("bsr/2 and >>>/2", () => {
    const bsr = Elixir_Bitwise["bsr/2"];
    const operator = Elixir_Bitwise[">>>/2"];

    it("shifts right for positive shift counts", () => {
      const result = bsr(Type.integer(9), Type.integer(2));
      const expected = Type.integer(2);

      assert.deepStrictEqual(result, expected);
    });

    it("converts negative shift counts to left shifts", () => {
      const result = bsr(Type.integer(1), Type.integer(-2));
      const expected = Type.integer(4);

      assert.deepStrictEqual(result, expected);
    });

    it("supports operator form", () => {
      const result = operator(Type.integer(-8), Type.integer(2));
      const expected = Type.integer(-8n >> 2n);

      assert.deepStrictEqual(result, expected);
    });
  });
});

File: test/elixir/hologram/ex_js_consistency/elixir/bitwise_test.exs
defmodule Hologram.ExJsConsistency.Elixir.BitwiseTest do
  @moduledoc """
  IMPORTANT!
  Each Elixir consistency test has a related JavaScript test in test/javascript/elixir/bitwise_test.mjs
  Always update both together.
  """

  use Hologram.Test.BasicCase, async: true

  @moduletag :consistency

  describe "band/2 and &&&/2" do
    test "calculates bitwise AND" do
      assert Bitwise.band(9, 3) == 1
    end

    test "handles negative operands" do
      assert Bitwise.&&&(-5, 12) == Bitwise.band(-5, 12)
    end

    test "raises ArithmeticError when an operand is not an integer" do
      assert_error ArithmeticError,
                   "bad argument in arithmetic expression: Bitwise.band(:foo, 1)",
                   {Bitwise, :band, [:foo, 1]}
    end
  end

  describe "bor/2 and |||/2" do
    test "calculates bitwise OR" do
      assert Bitwise.bor(9, 3) == 11
    end

    test "supports operator form" do
      assert Bitwise.|||(-5, 12) == Bitwise.bor(-5, 12)
    end
  end

  describe "bxor/2 and ^^^/2" do
    test "calculates bitwise XOR" do
      assert Bitwise.bxor(9, 3) == 10
    end

    test "supports operator form" do
      assert Bitwise.^^^(-5, 12) == Bitwise.bxor(-5, 12)
    end
  end

  describe "bnot/1 and ~~~/1" do
    test "calculates bitwise NOT" do
      assert Bitwise.bnot(2) == -3
    end

    test "supports operator form" do
      assert Bitwise.~~~(-1) == Bitwise.bnot(-1)
    end
  end

  describe "bsl/2 and <<</2" do
    test "shifts left for positive shift counts" do
      assert Bitwise.bsl(1, 3) == 8
    end

    test "converts negative shift counts to right shifts" do
      assert Bitwise.bsl(1, -2) == 0
    end

    test "supports operator form" do
      assert Bitwise.<<<(-1, 2) == Bitwise.bsl(-1, 2)
    end
  end

  describe "bsr/2 and >>>/2" do
    test "shifts right for positive shift counts" do
      assert Bitwise.bsr(9, 2) == 2
    end

    test "converts negative shift counts to left shifts" do
      assert Bitwise.bsr(1, -2) == 4
    end

    test "supports operator form" do
      assert Bitwise.>>>(-8, 2) == Bitwise.bsr(-8, 2)
    end
  end
end
